File: 0039-combination-sum/0039-combination-sum.java
class Solution {
    public void recSum(List<List<Integer>> res, int[] candidates, int target, List<Integer> curr, int start){
        if(target == 0){
            res.add(new ArrayList<>(curr));
            return;
        }
        if(target < 0) return;
        for(int i = start; i < candidates.length; i++){
            curr.add(candidates[i]);
            recSum(res, candidates, target - candidates[i], curr, i);
            curr.remove(curr.size()-1);
        }
    }
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> res = new ArrayList<>();
        recSum(res, candidates, target, new ArrayList<Integer>(), 0);
        return res;
    }
}

File: 0039-combination-sum/0039-combination-sum.py
class Solution:
    def combinationSum(self, nums, target):
        result = []
        nums.sort()
        self.backtrack(result, [], nums, target, 0)
        return result

    def backtrack(self, result, tempList, nums, remain, start):
        if remain < 0:
            return
        elif remain == 0:
            result.append(list(tempList))
        else:
            for i in range(start, len(nums)):
                tempList.append(nums[i])
                self.backtrack(result, tempList, nums, remain - nums[i], i)
                tempList.pop()


